provider/digitalocean/digital_ocean_test.go

```diff
@@ -143,10 +146,54 @@ func (m *mockDigitalOceanRecordsFail) Records(ctx context.Context, domain string
 	return []godo.DomainRecord{}, nil, fmt.Errorf("Failed to get records")
 }
 
-func TestNewDigitalOceanChanges(t *testing.T) {
-	action := DigitalOceanCreate
-	endpoints := []*endpoint.Endpoint{{DNSName: "new", Targets: endpoint.Targets{"target"}}}
-	_ = newDigitalOceanChanges(action, endpoints)
+// This function is an adapted copy of the testify package's ElementsMatch function with the
+// call to ObjectsAreEqual replaced with cmp.Equal which better handles struct's with pointers to
+// other structs.
+//
+// TODO: Upstream a change to testify package to allow custom matchers.
+func elementsMatch(t *testing.T, listA, listB interface{}, msgAndArgs ...interface{}) (ok bool) {
+	aKind := reflect.TypeOf(listA).Kind()
+	bKind := reflect.TypeOf(listB).Kind()
+
+	if aKind != reflect.Array && aKind != reflect.Slice {
+		return assert.Fail(t, fmt.Sprintf("%q has an unsupported type %s", listA, aKind), msgAndArgs...)
+	}
+
+	if bKind != reflect.Array && bKind != reflect.Slice {
+		return assert.Fail(t, fmt.Sprintf("%q has an unsupported type %s", listB, bKind), msgAndArgs...)
+	}
+
+	aValue := reflect.ValueOf(listA)
+	bValue := reflect.ValueOf(listB)
+
+	aLen := aValue.Len()
+	bLen := bValue.Len()
+
+	if aLen != bLen {
+		return assert.Fail(t, fmt.Sprintf("lengths don't match: %d != %d", aLen, bLen), msgAndArgs...)
+	}
+
+	// Mark indexes in bValue that we already used
+	visited := make([]bool, bLen)
+	for i := 0; i < aLen; i++ {
+		element := aValue.Index(i).Interface()
+		found := false
+		for j := 0; j < bLen; j++ {
+			if visited[j] {
+				continue
+			}
+			if cmp.Equal(bValue.Index(j).Interface(), element) {
+				visited[j] = true
+				found = true
+				break
+			}
+		}
+		if !found {
+			return assert.Fail(t, fmt.Sprintf("element %s appears more times in %s than in %s", element, aValue, bValue), msgAndArgs...)
+		}
+	}
+
+	return true
```

Comment:
Do we have test for this function? It's big enough for me to not be able to quickly tell if it works, so I'm worried that all the tests could be affected by a problem in this function. Do you think there is an easy way to make sure this function is correct?

Reply:
Good point. I will add some basic tests for the `elementsMatch` function.